lib/azure/armrest/virtual_machine_image_manager.rb

```diff
@@ -0,0 +1,135 @@
+# Azure namespace
+module Azure
+  # Armrest namespace
+  module Armrest
+    # Base class for managing virtual machine images
+    class VirtualMachineImageManager < ArmrestManager
+      # The location used in requests when gathering VM image information.
+      attr_accessor :location
+
+      # The provider used in requests when gathering VM image information.
+      attr_reader :provider
+
+      # The publisher used in requests when gathering VM image information.
+      attr_accessor :publisher
+
+      # Create and return a new VirtualMachineImageManager (VMIM) instance.
+      #
+      # This subclass accepts the additional :location, :provider, and
+      # :publisher options as well. The default provider is set to
+      # 'Microsoft.Compute'.
+      #
+      def initialize(options = {})
+        super
+
+        @location  = options[:location]
+        @provider  = options[:provider] || 'Microsoft.Compute'
+        @publisher = options[:publisher]
+
+        # Typically only empty in testing.
+        unless @@providers.empty?
+          @api_version = @@providers[@provider]['locations/publishers']['api_version']
+        end
+      end
+
+      # Set a new provider to use the default for other methods. This may alter
+      # the api_version used for future requests. In practice, only
+      # 'Microsoft.Compute' or 'Microsoft.ClassicCompute' should be used.
+      #
+      def provider=(name)
+        @api_version = @@providers[name]['locations/publishers']['api_version']
+        @provider = name
+      end
+
+      # Return a list of VM image offers from the given +publisher+ and +location+.
+      #
+      # Example:
+      #
+      #   vmim.offers('eastus', 'Canonical')
+      #
+      def offers(location = @location, publisher = @publisher)
+        raise ArgumentError, "No location specified" unless location
+        raise ArgumentError, "No publisher specified" unless publisher
+
+        url = build_url(location, 'publishers', publisher, 'artifacttypes', 'vmimage', 'offers')
+
+        JSON.parse(rest_get(url)).map{ |element| element['name'] }
```

Comment:
@bzwei, no, I double checked. There's no 'value' key in the returned json results. In the offers method, for example, the result of JSON.parse(rest_get(url)) is:

```[{"location"=>"eastus", "name"=>"Ubuntu15.04Snappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04Snappy"}, {"location"=>"eastus", "name"=>"Ubuntu15.04SnappyDocker", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04SnappyDocker"}, {"location"=>"eastus", "name"=>"UbunturollingSnappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbunturollingSnappy"}, {"location"=>"eastus", "name"=>"UbuntuServer", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbuntuServer"}]
[{"location"=>"eastus", "name"=>"Ubuntu15.04Snappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04Snappy"}, {"location"=>"eastus", "name"=>"Ubuntu15.04SnappyDocker", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/Ubuntu15.04SnappyDocker"}, {"location"=>"eastus", "name"=>"UbunturollingSnappy", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbunturollingSnappy"}, {"location"=>"eastus", "name"=>"UbuntuServer", "id"=>"/Subscriptions/45bf5648-d1f6-4dae-a6bd-52be617f5160/Providers/Microsoft.Compute/Locations/eastus/Publishers/Canonical/ArtifactTypes/VMImage/Offers/UbuntuServer"}]```